optuna/multi_objective/samplers/_nsga2.py

```diff
@@ -0,0 +1,256 @@
+_PARENTS_KEY = "multi_objective:nsga2:parents"
+
+
+@experimental("1.4.0")
+class NSGAIIMultiObjectiveSampler(BaseMultiObjectiveSampler):
+    """Multi-objective sampler using NSGA-II algorithm.
+
+    NSGA-II stands for "Nondominated Sorting Genetic Algorithm II",
+    which is a well known, fast and elitist multi-objective genetic algorithm.
+
+    For further information about NSGA-II, please refer to the following paper:
+
+    - `A fast and elitist multiobjective genetic algorithm: NSGA-II
+      <https://ieeexplore.ieee.org/document/996017>`_
+
+    Args:
+        seed:
+            Seed for random number generator.
+
+        population_size:
+            Number of individuals (trials) in a generation.
+
+        mutation_prob:
+            Probability of mutating each parameter when creating a new individual.
+
+        crossover_prob:
+            Probability of swapping each parameter of the parents when creating a new individual.
+
+    """
+
+    def __init__(
+        self,
+        seed: Optional[int] = None,
+        population_size: int = 50,
+        mutation_prob: float = 0.1,
+        crossover_prob: float = 0.2,
+    ) -> None:
+        # TODO(ohta): Reconsider the default value of each parameter.
+
+        if not isinstance(population_size, int):
+            raise ValueError("`population_size` must be an integer value.")
+
+        if population_size < 2:
+            raise ValueError("`population_size` must be greater than or equal to 2.")
+
+        if not (0.0 <= mutation_prob <= 1.0):
+            raise ValueError("`mutation_prob` must be a float value within the range [0.0, 1.0].")
+
+        if not (0.0 <= crossover_prob <= 1.0):
+            raise ValueError("`crossover_prob` must be a float value within the range [0.0, 1.0].")
+
+        self._population_size = population_size
+        self._crossover_prob = crossover_prob
+        self._mutation_prob = mutation_prob
+        self._random_sampler = multi_objective.samplers.RandomMultiObjectiveSampler(seed=seed)
+        self._rng = np.random.RandomState(seed)
+
+    def infer_relative_search_space(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        trial: "multi_objective.trial.FrozenMultiObjectiveTrial",
+    ) -> Dict[str, BaseDistribution]:
+        return {}
+
+    def sample_relative(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        trial: "multi_objective.trial.FrozenMultiObjectiveTrial",
+        search_space: Dict[str, BaseDistribution],
+    ) -> Dict[str, Any]:
+        parent_generation, population = self._collect_parent_population(study)
+        trial_id = trial._trial_id
+
+        if len(population) == 0:
+            generation = 0
+            study._storage.set_trial_system_attr(trial_id, _GENERATION_KEY, generation)
+        else:
+            p0 = self._select_parent(study, population)
+            p1 = self._select_parent(study, population)
```

Comment:
For smaller populations, it might be somewhat likely that the same parent is chosen for `p0` and `p1`. How about resampling until they're unique?

Reply:
> How about resampling until they're unique?

IMHO, resampling can frequently occur if the problem is simple and the Pareto-front is narrow.
How about removing `p0` from `population` to sample `p1`?

Reply:
Thank you for your advice! I adopted the logic suggested by @toshihikoyanase at the commit https://github.com/optuna/optuna/pull/1163/commits/eac981f1092e324bfd9beb60638368b17d27ac3f. 
Besides https://github.com/optuna/optuna/pull/1163#discussion_r424243213 and https://github.com/optuna/optuna/pull/1163#discussion_r424242441 were also addressed in the same commit.